custom_components/nest_protect/__init__.py

```diff
@@ -116,39 +117,63 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
 
     devices: dict[str, Bucket] = {b.object_key: b for b in device_buckets}
 
-    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = HomeAssistantNestProtectData(
+    entry_data = HomeAssistantNestProtectData(
         devices=devices,
         areas=areas,
         client=client,
     )
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = entry_data
 
     await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
 
     # Subscribe for real-time updates
-    # TODO cancel when closing HA / unloading entry
-    _register_subscribe_task(hass, entry, data)
+    entry_data.subscription_task = asyncio.create_task(
+        _async_subscribe_for_data(hass, entry, data)
+    )
 
     return True
 
 
 async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Unload a config entry."""
+    # Cancel subscription task before unloading
+    if entry.entry_id in hass.data.get(DOMAIN, {}):
```

Comment:
Is this additional check required? We only remove/pop the entry id from hass.data on line 150 right?